File: scripts/convert_tei_to_json.py
#!/usr/bin/env python3
"""
Convert Rig Veda TEI file to JSON format matching rigveda.sample.json
"""

import xml.etree.ElementTree as ET
import json
import re
from typing import Dict, List, Any

def clean_text(text: str) -> str:
    """Clean and normalize text content"""
    if not text:
        return ""
    # Remove extra whitespace and normalize
    text = re.sub(r'\s+', ' ', text.strip())
    return text

def extract_verse_content(stanza_elem) -> Dict[str, str]:
    """Extract Sanskrit, transliteration, and translation from a stanza"""
    verse_data = {
        'sanskrit_deva': '',
        'transliteration': '',
        'translation': {
            'text': '',
            'source': 'R.T.H. Griffith, 1896',
            'license': 'Public Domain'
        }
    }
    
    # Find Sanskrit Devanagari text (eichler source)
    sanskrit_elem = stanza_elem.find('.//lg[@xml:lang="san-Deva"][@source="eichler"]', 
                                   {'xml': 'http://www.tei-c.org/ns/1.0'})
    if sanskrit_elem is not None:
        sanskrit_lines = []
        for l_elem in sanskrit_elem.findall('.//l', {'xml': 'http://www.tei-c.org/ns/1.0'}):
            line_text = clean_text(l_elem.text or '')
            if line_text:
                sanskrit_lines.append(line_text)
        verse_data['sanskrit_deva'] = ' '.join(sanskrit_lines)
    
    # Find transliteration (zurich source)
    translit_elem = stanza_elem.find('.//lg[@xml:lang="san-Latn-x-ISO-15919"][@source="zurich"]', 
                                   {'xml': 'http://www.tei-c.org/ns/1.0'})
    if translit_elem is not None:
        translit_lines = []
        for l_elem in translit_elem.findall('.//l', {'xml': 'http://www.tei-c.org/ns/1.0'}):
            line_text = clean_text(l_elem.text or '')
            if line_text:
                translit_lines.append(line_text)
        verse_data['transliteration'] = ' '.join(translit_lines)
    
    # Find English translation (griffith source)
    translation_elem = stanza_elem.find('.//lg[@xml:lang="eng"][@source="griffith"]', 
                                      {'xml': 'http://www.tei-c.org/ns/1.0'})
    if translation_elem is not None:
        translation_lines = []
        for l_elem in translation_elem.findall('.//l', {'xml': 'http://www.tei-c.org/ns/1.0'}):
            line_text = clean_text(l_elem.text or '')
            if line_text:
                translation_lines.append(line_text)
        verse_data['translation']['text'] = ' '.join(translation_lines)
    
    return verse_data

def extract_hymn_metadata(hymn_elem) -> Dict[str, str]:
    """Extract hymn metadata (deity, rishi, meter)"""
    metadata = {
        'deity': 'Unknown',
        'rsi': 'Unknown', 
        'meter': 'Unknown'
    }
    
    # Find dedication section
    dedication = hymn_elem.find('.//div[@type="dedication"]', {'xml': 'http://www.tei-c.org/ns/1.0'})
    if dedication is not None:
        # Extract deity
        addressee = dedication.find('.//div[@type="addressee"]', {'xml': 'http://www.tei-c.org/ns/1.0'})
        if addressee is not None:
            eng_p = addressee.find('.//p[@xml:lang="eng"]', {'xml': 'http://www.tei-c.org/ns/1.0'})
            if eng_p is not None and eng_p.text:
                metadata['deity'] = clean_text(eng_p.text)
        
        # Extract rishi
        group = dedication.find('.//div[@type="group"]', {'xml': 'http://www.tei-c.org/ns/1.0'})
        if group is not None:
            eng_p = group.find('.//p[@xml:lang="eng"]', {'xml': 'http://www.tei-c.org/ns/1.0'})
            if eng_p is not None and eng_p.text:
                # Extract rishi name from "Hymns of [Rishi]"
                rishi_text = clean_text(eng_p.text)
                if rishi_text.startswith('Hymns of '):
                    metadata['rsi'] = rishi_text.replace('Hymns of ', '')
                else:
                    metadata['rsi'] = rishi_text
    
    return metadata

def convert_tei_to_json(tei_file_path: str, output_file_path: str):
    """Convert TEI file to JSON format"""
    print(f"Parsing TEI file: {tei_file_path}")
    
    # Parse the XML file
    tree = ET.parse(tei_file_path)
    root = tree.getroot()
    
    # Define namespace
    ns = {'xml': 'http://www.tei-c.org/ns/1.0'}
    
    # Find the book element
    book_elem = root.find('.//div[@type="book"]', ns)
    if book_elem is None:
        raise ValueError("Could not find book element in TEI file")
    
    # Extract mandala number
    mandala_num = int(book_elem.get('xml:id', '1').replace('b', ''))
    
    print(f"Processing Mandala {mandala_num}")
    
    # Process hymns
    hymns = []
    hymn_elements = book_elem.findall('.//div[@type="hymn"]', ns)
    
    print(f"Found {len(hymn_elements)} hymns")
    
    for i, hymn_elem in enumerate(hymn_elements):
        hymn_id = hymn_elem.get('xml:id', f'b{mandala_num:02d}_h{i+1:03d}')
        hymn_num = int(hymn_elem.get('ana', str(i+1)))
        
        print(f"Processing hymn {hymn_num}...")
        
        # Extract metadata
        metadata = extract_hymn_metadata(hymn_elem)
        
        # Process verses (stanzas)
        verses = []
        stanza_elements = hymn_elem.findall('.//div[@type="stanza"]', ns)
        
        for j, stanza_elem in enumerate(stanza_elements):
            verse_data = extract_verse_content(stanza_elem)
            
            # Only include verses that have content
            if (verse_data['sanskrit_deva'] or 
                verse_data['transliteration'] or 
                verse_data['translation']['text']):
                
                verses.append({
                    'verseNumber': j + 1,
                    'sanskrit_deva': verse_data['sanskrit_deva'],
                    'transliteration': verse_data['transliteration'],
                    'translation': verse_data['translation']
                })
        
        # Only include hymns that have verses
        if verses:
            hymns.append({
                'suktaId': f'{mandala_num}-{hymn_num}',
                'suktaNumber': hymn_num,
                'rsi': metadata['rsi'],
                'deity': metadata['deity'],
                'meter': metadata['meter'],
                'verses': verses
            })
    
    # Create the final JSON structure
    result = {
        'mandalas': [{
            'mandala': mandala_num,
            'suktas': hymns
        }]
    }
    
    # Write to output file
    print(f"Writing to {output_file_path}")
    with open(output_file_path, 'w', encoding='utf-8') as f:
        json.dump(result, f, ensure_ascii=False, indent=2)
    
    print(f"Conversion complete!")
    print(f"Generated {len(hymns)} hymns with {sum(len(h['verses']) for h in hymns)} verses")

if __name__ == '__main__':
    import sys
    
    if len(sys.argv) != 3:
        print("Usage: python convert_tei_to_json.py <input_tei_file> <output_json_file>")
        sys.exit(1)
    
    input_file = sys.argv[1]
    output_file = sys.argv[2]
    
    try:
        convert_tei_to_json(input_file, output_file)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

File: scripts/convert_tei_to_json.js
#!/usr/bin/env node
/**
 * Convert Rig Veda TEI file to JSON format matching rigveda.sample.json
 */

const fs = require('fs');
const path = require('path');

// Simple XML parser for our specific use case
function parseXML(xmlString) {
  const lines = xmlString.split('\n');
  const result = { mandalas: [] };
  let currentMandala = null;
  let currentHymn = null;
  let currentStanza = null;
  let currentVerse = 0;
  
  let mandalaNum = 1;
  let hymnNum = 1;
  let verseNum = 1;
  
  const hymns = [];
  
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    
    // Extract mandala number
    if (line.includes('xml:id="b01" type="book"')) {
      mandalaNum = 1;
    }
    
    // Extract hymn number and metadata
    if (line.includes('type="hymn"')) {
      const hymnMatch = line.match(/ana="(\d+)"/);
      if (hymnMatch) {
        hymnNum = parseInt(hymnMatch[1]);
      }
      
      // Reset for new hymn
      currentHymn = {
        suktaId: `${mandalaNum}-${hymnNum}`,
        suktaNumber: hymnNum,
        rsi: 'Unknown',
        deity: 'Unknown',
        meter: 'Unknown',
        verses: []
      };
      currentVerse = 0;
    }
    
    // Extract deity
    if (line.includes('xml:lang="eng"') && line.includes('Agni')) {
      currentHymn.deity = 'Agni';
    } else if (line.includes('xml:lang="eng"') && line.includes('Indra')) {
      currentHymn.deity = 'Indra';
    } else if (line.includes('xml:lang="eng"') && line.includes('Soma')) {
      currentHymn.deity = 'Soma';
    } else if (line.includes('xml:lang="eng"') && line.includes('Varuna')) {
      currentHymn.deity = 'Varuna';
    } else if (line.includes('xml:lang="eng"') && line.includes('Vishnu')) {
      currentHymn.deity = 'Vishnu';
    } else if (line.includes('xml:lang="eng"') && line.includes('Mitra')) {
      currentHymn.deity = 'Mitra';
    } else if (line.includes('xml:lang="eng"') && line.includes('Ushas')) {
      currentHymn.deity = 'Ushas';
    } else if (line.includes('xml:lang="eng"') && line.includes('Rudra')) {
      currentHymn.deity = 'Rudra';
    } else if (line.includes('xml:lang="eng"') && line.includes('Maruts')) {
      currentHymn.deity = 'Maruts';
    } else if (line.includes('xml:lang="eng"') && line.includes('Ashvins')) {
      currentHymn.deity = 'Ashvins';
    }
    
    // Extract rishi
    if (line.includes('Hymns of')) {
      const rishiMatch = line.match(/Hymns of ([^<]+)/);
      if (rishiMatch) {
        currentHymn.rsi = rishiMatch[1].trim();
      }
    }
    
    // Extract Sanskrit Devanagari text
    if (line.includes('xml:lang="san-Deva"') && line.includes('source="eichler"')) {
      // Look for the next few lines for Sanskrit text
      for (let j = i + 1; j < i + 5 && j < lines.length; j++) {
        const nextLine = lines[j].trim();
        if (nextLine.startsWith('<l xml:id=') && nextLine.includes('eichler')) {
          const sanskritMatch = nextLine.match(/>([^<]+)</);
          if (sanskritMatch) {
            if (!currentHymn.verses[currentVerse]) {
              currentHymn.verses[currentVerse] = {
                verseNumber: currentVerse + 1,
                sanskrit_deva: '',
                transliteration: '',
                translation: {
                  text: '',
                  source: 'R.T.H. Griffith, 1896',
                  license: 'Public Domain'
                }
              };
            }
            currentHymn.verses[currentVerse].sanskrit_deva += sanskritMatch[1] + ' ';
          }
        }
      }
    }
    
    // Extract transliteration
    if (line.includes('xml:lang="san-Latn-x-ISO-15919"') && line.includes('source="zurich"')) {
      // Look for the next few lines for transliteration
      for (let j = i + 1; j < i + 5 && j < lines.length; j++) {
        const nextLine = lines[j].trim();
        if (nextLine.startsWith('<l xml:id=') && nextLine.includes('zur')) {
          const translitMatch = nextLine.match(/>([^<]+)</);
          if (translitMatch) {
            if (!currentHymn.verses[currentVerse]) {
              currentHymn.verses[currentVerse] = {
                verseNumber: currentVerse + 1,
                sanskrit_deva: '',
                transliteration: '',
                translation: {
                  text: '',
                  source: 'R.T.H. Griffith, 1896',
                  license: 'Public Domain'
                }
              };
            }
            currentHymn.verses[currentVerse].transliteration += translitMatch[1] + ' ';
          }
        }
      }
    }
    
    // Extract English translation
    if (line.includes('xml:lang="eng"') && line.includes('source="griffith"')) {
      // Look for the next few lines for translation
      for (let j = i + 1; j < i + 5 && j < lines.length; j++) {
        const nextLine = lines[j].trim();
        if (nextLine.startsWith('<l xml:id=') && nextLine.includes('griffith')) {
          const transMatch = nextLine.match(/>([^<]+)</);
          if (transMatch) {
            if (!currentHymn.verses[currentVerse]) {
              currentHymn.verses[currentVerse] = {
                verseNumber: currentVerse + 1,
                sanskrit_deva: '',
                transliteration: '',
                translation: {
                  text: '',
                  source: 'R.T.H. Griffith, 1896',
                  license: 'Public Domain'
                }
              };
            }
            currentHymn.verses[currentVerse].translation.text += transMatch[1] + ' ';
          }
        }
      }
    }
    
    // Detect new stanza/verse
    if (line.includes('type="stanza"')) {
      currentVerse++;
    }
    
    // End of hymn - add to results
    if (line.includes('</div>') && currentHymn && currentHymn.verses.length > 0) {
      // Clean up the verses
      currentHymn.verses = currentHymn.verses.filter(v => v && (v.sanskrit_deva.trim() || v.transliteration.trim() || v.translation.text.trim()));
      
      if (currentHymn.verses.length > 0) {
        hymns.push(currentHymn);
      }
      currentHymn = null;
    }
  }
  
  result.mandalas = [{
    mandala: mandalaNum,
    suktas: hymns
  }];
  
  return result;
}

function cleanText(text) {
  if (!text) return '';
  return text.replace(/\s+/g, ' ').trim();
}

function convertTEIToJSON(inputFile, outputFile) {
  console.log(`Reading TEI file: ${inputFile}`);
  
  const xmlContent = fs.readFileSync(inputFile, 'utf8');
  console.log(`File size: ${xmlContent.length} characters`);
  
  console.log('Parsing XML...');
  const result = parseXML(xmlContent);
  
  console.log(`Found ${result.mandalas[0].suktas.length} hymns`);
  
  // Clean up the data
  result.mandalas[0].suktas.forEach(hymn => {
    hymn.verses.forEach(verse => {
      verse.sanskrit_deva = cleanText(verse.sanskrit_deva);
      verse.transliteration = cleanText(verse.transliteration);
      verse.translation.text = cleanText(verse.translation.text);
    });
  });
  
  console.log(`Writing to ${outputFile}`);
  fs.writeFileSync(outputFile, JSON.stringify(result, null, 2), 'utf8');
  
  console.log('Conversion complete!');
}

// Main execution
const args = process.argv.slice(2);
if (args.length !== 2) {
  console.log('Usage: node convert_tei_to_json.js <input_tei_file> <output_json_file>');
  process.exit(1);
}

const inputFile = args[0];
const outputFile = args[1];

if (!fs.existsSync(inputFile)) {
  console.error(`Input file not found: ${inputFile}`);
  process.exit(1);
}

try {
  convertTEIToJSON(inputFile, outputFile);
} catch (error) {
  console.error('Error:', error.message);
  process.exit(1);
}





